src/Servers/Kestrel/Transport.MsQuic/Internal/MsQuicConnectionListener.cs

```diff
@@ -0,0 +1,160 @@
+            {
+                return _acceptEnumerator.Current;
+            }
+
+            return null;
+        }
+
+        public async ValueTask DisposeAsync()
+        {
+            if (_disposed)
+            {
+                return;
+            }
+
+            _disposed = true;
+
+            StopAcceptingConnections();
+
+            await UnbindAsync().ConfigureAwait(false);
+
+            // TODO do something with _listener;
+
+            //await StopThreadsAsync().ConfigureAwait(false);
+        }
+
+        public async ValueTask UnbindAsync(CancellationToken cancellationToken = default)
+        {
+            if (_stopped)
+            {
+                return;
+            }
+
+            _stopped = true;
+
+            await DisposeAsync();
+        }
+
+        internal async Task BindAsync()
+        {
+            // TODO make this configurable
+            await StartAsync();
+
+            _acceptEnumerator = AcceptConnectionsAsync();
+        }
+
+        private async IAsyncEnumerator<MsQuicConnection> AcceptConnectionsAsync()
+        {
+            while (true)
+            {
+                while (await _acceptConnectionQueue.Reader.WaitToReadAsync())
+                {
+                    while (_acceptConnectionQueue.Reader.TryRead(out var connection))
+                    {
+                        yield return connection;
+                    }
+                }
+
+                yield return null;
+            }
+        }
+
+        public async Task StartAsync(CancellationToken cancellationToken = default)
+        {
+            _registration.RegistrationOpen(Encoding.ASCII.GetBytes(TransportContext.Options.RegistrationName));
+
+            _secConfig = await _registration.CreateSecurityConfig(TransportContext.Options.Certificate);
+
+            _session = _registration.SessionOpen(TransportContext.Options.Alpn);
+
+            _listener = _session.ListenerOpen(ListenerCallbackHandler);
+            _session.SetIdleTimeout(TransportContext.Options.IdleTimeout);
+            _session.SetPeerBiDirectionalStreamCount(TransportContext.Options.MaxBidirectionalStreamCount);
+            _session.SetPeerUnidirectionalStreamCount(TransportContext.Options.MaxBidirectionalStreamCount);
+
+            _listener.Start(EndPoint as IPEndPoint);
+        }
+
+        private QUIC_STATUS ListenerCallbackHandler(
+            ref NativeMethods.ListenerEvent evt)
+        {
+            switch (evt.Type)
+            {
+                case QUIC_LISTENER_EVENT.NEW_CONNECTION:
+                    {
+                        var connection = new QuicConnection(_registration, evt.Data.NewConnection.Connection, false);
+
+                        evt.Data.NewConnection.SecurityConfig = _secConfig.NativeObjPtr;
+                        var msQuicConnection = new MsQuicConnection(connection);
+                        _acceptConnectionQueue.Writer.TryWrite(msQuicConnection);
```

Comment:
Once you return success from this (and not pending) then the connection is going to continue on, and you can start getting callbacks, even streams being open and callbacks on those. If you want to do this async accept design, then you should return pending, and only when accepted, do you SetParam(SecConfig) to completely accept the connection.

Reply:
I'd mark this as a TODO as well.